rfcs/plugins.md

```diff
@@ -0,0 +1,349 @@
+    'hello-world': (req, res, state) => {
+        res.send(`Hello ${state.value}`);
+            res.end()
+    },
+}
+```
+
+Notes:
+ - The handler is responsible for calling `res.end()`.
+ - The route is a hard-prefix from the root of the horizon server address.
+  - It cannot clash with any other other loaded plugin routes (no
+    overlapping)
+      - If multiple routes need to be handled, the r
+      - Example, if you want to create a REST plugin, you need to
+        specify the route as something like `rest` and the endpoints
+        underneath it like `/rest/users` `/rest/vehicles` etc.
+  - It cannot clash with the route horizon uses itself (usually
+    `"horizon"`)
+ - In the configuration, the user may override the httpRoute for the
+   plugin, and the plugin must behave properly.
+ - Express 4.0 api for request and response objects
+
+### Processes
+
+The `processes` plugin property is an object with keys that correspond
+to process titles, and values which are objects with 2 properties:
+ - `run`: a function receiving the plugin state that runs as the main
+   thread of the process. It should not spawn processes itself
+ - `cleanup`: a function receiving the plugin state that cleans up the
+   process
+
+### Environment vars
+
+The `env` plugin property is an object from keys with environment
+variables, and Joi validators for values.
+
+Example:
+```js
+env: {
+  'MY_ENV_VALUE': Joi.any().allow('yes', 'no').default('no')
+}
+```
+
+The values of env variables that successfully validate are passed to
+the [activate](#activate) function as well as to the `processConfig` function for any [command](#command).
+
+**Note**:
+
+The env vars declared here will be validated to ensure no other plugin
+declares them. If a plugin wants to make use of externally defined env
+variables like `$PATH` etc, and doesn't want to mutual exclusivity to be enforced by Horizon, it can just access `process.env` like normal.
+
+### Validation helpers
+
+The `validationHelpers` plugin property is an object containing
+functions that will be made available in the validation context. There
+are no restrictions on the types of the functions here
+
+Example:
+```js
+validationHelpers: {
+ isTruthy(val) { return !!val }
+}
+```
+
+This will make the `isTruthy` function available to the `validators`
+functions specified in `.hz/config.toml`:
+
+```toml
+[groups.authenticated.rules.foo]
+validator = """
+  function(context, oldValue, newValue) {
+    return isTruthy(newValue.isFoo)
+  }
+"""
+```
+
+### Requests
+
+The `requests` plugin property allows specifying new request types
+that can be sent by the client. The keys of the provided object are
+the name of the new requests, the values are an object with the
+following properties:
+  - `optionsSchema`: A Joi schema to validate the request type.
+  - `clientValidation`: An object with the keys:
+    - `minArgs`: minimum number of arguments accepted
+    - `maxArgs`: maximum number of arguments accepted
```

Comment:
Nope, and none of the built in commands allow an unlimited number either (they just use 100 as the max). This is arguably stupid, but I think it simplified some code somewhere